sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiLoads.java

```diff
@@ -154,43 +154,61 @@ public WriteResult expandTriggered(
                     successfulRowsTag,
                     BigQueryStorageApiInsertErrorCoder.of(),
                     successCoder));
-    PCollection<KV<ShardedKey<DestinationT>, StorageApiWritePayload>> shardedRecords =
-        result
-            .get(successfulRowsTag)
-            .apply(
-                "AddShard",
-                ParDo.of(
-                    new DoFn<
-                        KV<DestinationT, StorageApiWritePayload>,
-                        KV<ShardedKey<DestinationT>, StorageApiWritePayload>>() {
-                      int shardNumber;
-
-                      @Setup
-                      public void setup() {
-                        shardNumber = ThreadLocalRandom.current().nextInt(numShards);
-                      }
-
-                      @ProcessElement
-                      public void processElement(
-                          @Element KV<DestinationT, StorageApiWritePayload> element,
-                          OutputReceiver<KV<ShardedKey<DestinationT>, StorageApiWritePayload>> o) {
-                        DestinationT destination = element.getKey();
-                        ByteBuffer buffer = ByteBuffer.allocate(Integer.BYTES);
-                        buffer.putInt(++shardNumber % numShards);
-                        o.output(
-                            KV.of(ShardedKey.of(destination, buffer.array()), element.getValue()));
-                      }
-                    }))
-            .setCoder(KvCoder.of(ShardedKey.Coder.of(destinationCoder), payloadCoder));
-
-    PCollection<KV<ShardedKey<DestinationT>, Iterable<StorageApiWritePayload>>> groupedRecords =
-        shardedRecords.apply(
-            "GroupIntoBatches",
-            GroupIntoBatches.<ShardedKey<DestinationT>, StorageApiWritePayload>ofByteSize(
-                    MAX_BATCH_SIZE_BYTES,
-                    (StorageApiWritePayload e) -> (long) e.getPayload().length)
-                .withMaxBufferingDuration(triggeringFrequency));
 
+    PCollection<KV<ShardedKey<DestinationT>, Iterable<StorageApiWritePayload>>> groupedRecords;
+
+    if (this.allowAutosharding) {
+      groupedRecords =
+          result
+              .get(successfulRowsTag)
+              .apply(
+                  "GroupIntoBatches",
+                  GroupIntoBatches.<DestinationT, StorageApiWritePayload>ofByteSize(
+                          MAX_BATCH_SIZE_BYTES,
+                          (StorageApiWritePayload e) -> (long) e.getPayload().length)
+                      .withMaxBufferingDuration(triggeringFrequency)
+                      .withShardedKey());
+
+    } else {
+      PCollection<KV<ShardedKey<DestinationT>, StorageApiWritePayload>> shardedRecords =
```

Comment:
can we push this to a helper method to improve readability?

Reply:
Refactored, PTAL.